sway/desktop/render.c

```diff
@@ -1780,34 +1908,55 @@ static void render_floating_container(struct sway_output *soutput,
 			.discard_transparent = false,
 			.shadow = con->shadow_enabled,
 		};
-		render_view(soutput, damage, con, colors, deco_data);
+		render_view(soutput, damage, con, colors, deco_data, is_focused_ws);
 		if (has_titlebar) {
 			render_titlebar(soutput, damage, con, floor(state->x), floor(state->y),
 					state->width, colors, deco_data.alpha, deco_data.corner_radius,
-					ALL, title_texture, marks_texture);
+					ALL, title_texture, marks_texture, is_focused_ws);
 		} else if (state->border == B_PIXEL) {
-			render_top_border(soutput, damage, state, colors, deco_data.alpha, deco_data.corner_radius);
+			render_top_border(soutput, damage, state, colors, deco_data.alpha,
+					deco_data.corner_radius, is_focused_ws);
 		}
 	} else {
-		render_container(soutput, damage, con, state->focused);
+		render_container(soutput, damage, con, state->focused, is_focused_ws);
 	}
 }
 
 static void render_floating(struct sway_output *soutput,
-		pixman_region32_t *damage) {
+		pixman_region32_t *damage, struct sway_workspace *other_ws,
+		bool has_fullscreen) {
 	for (int i = 0; i < root->outputs->length; ++i) {
 		struct sway_output *output = root->outputs->items[i];
+
+		// Don't render floating windows across outputs when switching workspaces
+		if (output->workspace_scroll.percent != 0 && output != soutput) {
+			continue;
+		}
+
+		struct sway_workspace *visible_ws = output->current.active_workspace;
 		for (int j = 0; j < output->current.workspaces->length; ++j) {
 			struct sway_workspace *ws = output->current.workspaces->items[j];
-			if (!workspace_is_visible(ws)) {
+
+			float scroll_percent = soutput->workspace_scroll.percent;
+
+			// Only render visible workspace when not scrolling
+			bool ws_is_visible = workspace_is_visible(ws);
+			if (!ws_is_visible && scroll_percent == 0) {
+				continue;
+			}
```

Comment:
under which conditions would this logic be hit? I'd think if we arent scrolling we'd always be rendering visible ws only